libs/reconciler/README.md

```diff
@@ -0,0 +1,475 @@
+# State Reconciler
+
+## Problem Statement
+
+Configuration management agents typically receive a declarative description
+of the intended (also called desired) state to apply into the managed system.
+This description is usually at a higher level of abstraction and expresses
+the intent, not the implementation. The agent then *renders* a corresponding
+set of low-level run-time stateful objects (e.g. processes to run, files to write,
+config primitives to create), that, once *applied* into the running system,
+will implement the requested intent. The process of updating the system state is
+not trivial, however. The agent first have to compare the new intended state with
+the actual (also known as the current) state of the system. Based on the difference,
+the agent determines the sequence of operations (state transitions) to execute,
+in order to synchronize the system state with the new intent. The order of operations
+often matters as it has to respect all dependencies that exist between configured objects.
+Operations are usually described using the CRUD interface (Create, Update/Modify, Read
+and Delete), but a different semantic for state manipulation is also possible.
+
+To summarize, the problem described above can be split into the following tasks:
+
+  1. From a high-level declarative description (received from a user/controller),
+     render the corresponding set of configuration primitives, processes or files
+     to create/run/write in the running system.
+  2. Determine the difference between the current system state and the new (rendered)
+     intended state.
+  3. Determine the set of (Create/Modify/Delete) operations to execute and find a valid
+     ordering.
+  4. Execute operations, collect any errors and report back to the user/controller.
+
+While the task no. 1 is specific to each management agent, tasks 2, 3 and to some
+extent also 4 can be tackled in a generic way and in one place using **Reconciler**
+provided by this package.
+
+## Solution
+
+Reconciler is able to determine, properly order and drive state transitions needed
+to execute to reach the intended state. It requires that the current state as well as
+the intended state are each modeled as a separate instance of a [dependency graph](https://en.wikipedia.org/wiki/Dependency_graph),
+provided by the [depgraph package](../depgraph/README.md).
+The idea behind using a dependency graph for state representation is described in detail
+in the depgraph package. To summarize: every configuration primitive (OS config, file,
+process, etc.), further denoted as *configuration item* or just *item*, can be represented
+by a single graph node, while dependencies between items are modeled using directed edges.
+The graph supports dependency with a semantics *"must exist"*.\
```

Comment:
Extra `\` at the end of the line.

Reply:
This is forced line-break: https://szymonkrajewski.pl/line-break-in-markdown/